design-documents/graph-ql/coverage/storefront-route.md

```diff
@@ -0,0 +1,292 @@
+
+## Requirements
+
+As stated above, the PWA must be able to route:
+
+1) SEO URLs for catalog and CMS pages
+2) Built-in Magento routes, like `/checkout`
+3) Custom routes installed by extensions and other modules
+
+The current PWA implementation has a partial solution for **(1)**, using the [`urlResolver`](https://github.com/magento/graphql-ce/issues/80).
+It has a suboptimal solution for **(2)**, in the form of reimplementing Magento routing logic in the client-side app router.
+There is no current solution for **(3)**.
+
+To support Magento's SEO and display customization features, the PWA must be able to query the API for the following data:
+
+- **Entity type**, page type or content type (e.g. category, product, or CMS page)
+- **Entity** content and data
+
+With that information, the PWA can asynchronously load the view components which handle page display for the appropriate entity.
+
+Additionally, the PWA should be able to query the API for **display metadata** for any entity.
+This could be a simple string representing a shared "variant" component label, which would enable PWAs to support the equivalent of per-entity Theme Variants in existing Magento 2.
+
+<!-- diagram: url => { entity_type, entity, metadata } -->
+
+The current `urlResolver` query delivers part of these requirements: entity type in the form of an enum of "magic strings", such as `CATEGORY` and `CMS_PAGE`, and entity data in the form of a unique ID, which functions as a foreign key to be used in a subsequent GraphQL call.
+
+<!-- diagram: url => { type, id } -->
+
+This solution helps us honor SEO URLs, which is a _distinguishing feature_ of PWA Studio.
+However, it also has flaws:
+
+- The entity type enum must be hand-maintained and mapped to real types
+- Not all routable entities have implemented types: only `CATEGORY`, `CMS_PAGE` and `PRODUCT`.
+- The associated entity must be fetched in a subsequent API call
+- No metadata exists for page display strategy
+- Page types are not discoverable or extensible
+
+## Proposal: GraphQL `RoutableInterface`
+
+Add a `RoutableInterface` to the base schema, and require all entity types which are routable to implement it.
+
+A subset of Magento entities are "routable", in that they have URLs and they serve as the model for a rendered page.
+The natural fit for this concept is an interface, and GraphQL provides [multiple interfaces](https://graphql.github.io/graphql-spec/June2018/#sec-Objects) so that a given type can implement more than one.
+
+As described in [Requirements](#requirements), the PWA must have the entity type and entity data for a given route, and it should be able to query for display metadata for that route. A GraphQL interface is a good match for these requirements:
+
+- Get the entity type using `__typename` introspection to get the GraphQL concrete type of the `RoutableInterface`
```

Comment:
`__typename` may be disabled if the merchant decides to disable introspection (it should be now possible via `config.php` in Magento)